### PR TITLE
Introduce `TyErr` independent from `TyInfer`

### LINKED ISSUE #34264: bogus parameter-count error following error in parameter list
``` rust
fn foo(Option<i32>) {}

fn main() {
    foo(Some(42));
}
```

The first error is about the missing argument name in `foo`'s parameter list, but there's another one on the call saying that the function doesn't take any parameters. I guess a dummy item was created to prevent cascading resolve errors, but it seems to create another kind of cascading error.

### PR DESCRIPTION
**Pull request by estebank**
Add a `TyErr` type to represent unknown types in places where
parse errors have happened, while still able to build the AST.

Initially only used to represent incorrectly written fn arguments and
avoid "expected X parameters, found Y" errors when called with the
appropriate amount of parameters. We cannot use `TyInfer` for this as
`_` is not allowed as a valid argument type.

Example output:

```rust
error: expected one of `:` or `@`, found `,`
  --> file.rs:12:9
   |
12 | fn bar(x, y: usize) {}
   |         ^

error[E0061]: this function takes 2 parameters but 3 parameters were supplied
  --> file.rs:19:9
   |
12 | fn bar(x, y) {}
   | --------------- defined here
...
19 |     bar(1, 2, 3);
   |         ^^^^^^^ expected 2 parameters
```

Fix #34264.

### CODE REVIEW

**Comment by petrochenkov** [comment]
`TyError` would be a more appropriate name (by analogy with `TypeVariants::TyError` or `Def::Err` doing the similar job).

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4330,6 +4330,26 @@ impl<'a> Parser<'a> {
 
     fn parse_fn_args(&mut self, named_args: bool, allow_variadic: bool)
                      -> PResult<'a, (Vec<Arg> , bool)> {
+
+        /// Create a placeholder argument
+        fn dummy_arg(span: Span) -> Arg {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Nit: error reporting is a "cold" code for a reader trying to understand the main logic (syntax of fn arguments in this case), it often makes sense to factor it into a separate function (already done) and move it somewhere else, so it's invisible unless intentionally looked for.

Comment (estebank) [inline src/libsyntax/parse/parser.rs]: Done.

**Comment by estebank** [comment]
@petrochenkov done.

**Comment by estebank** [comment]
@bors r=petrochenkov
fixed merge conflicts

**Comment by TimNN** [comment]
@bors retry

#41147

**Pull request merged**